vowpalwabbit/crossplat_compat.h

```diff
@@ -4,14 +4,50 @@
 #pragma once
 
 #include <cstdint>
+#include <cstring>
+#include <string.h>
+#include <stdio.h>
+#include <cstdio>
 
 #ifndef _WIN32
 #define sprintf_s snprintf
 #define vsprintf_s vsnprintf
+#define strtok_s strtok_r
+#define fscanf_s fscanf
 
 constexpr uint64_t UINT64_ZERO = 0ULL;
 constexpr uint64_t UINT64_ONE = 1ULL;
 #else
 constexpr uint64_t UINT64_ONE = 1i64;
 constexpr uint64_t UINT64_32ONES = 0x00000000ffffffffi64;
 #endif
+
+namespace VW
+{
+inline int string_cpy(char *dest, size_t dest_size, const char *src)
+{
+#ifdef _WIN32
+  // strcpy_s returns an errno_t
+  return strcpy_s(dest, dest_size, src);
+#else
+  (void)dest_size; // unused here
+  strcpy(dest, src);
+  return 0;
+#endif
+}
+
+inline int file_open(FILE **pf, const char *filename, const char *mode)
```

Comment:
Should this new function be used in the file io adapters?
https://github.com/VowpalWabbit/vowpal_wabbit/blob/d4476ff327dc9416915b456b78d17da032c66a44/vowpalwabbit/io/io_adapter.cc#L241

Reply:
It looks like the io_adapter uses `_sopen_s/open` to get a file descriptor and do a lower level read/write of the underlying file vs `fopen_s/fopen` which provide for example buffering, so I am not sure they are suitable here?

Reply:
Got it, no prob